Linear quantization for signed integers

pabvald opened the pull request:
@milankl 
I have the basic implementation:
```julia 
# minimum and maximum representable value of type T
Tmin, Tmax = Float64(typemin(T)), Float64(typemax(T))

# LinQuantization
Δ⁻¹ =  (Tmax-Tmin)/(Amax-Amin) # (Tmax -  Tmin) == (2^sizeof(T)-1), but  imo Tmax - Tmin is more informative and the values are used afterwards again anyways

Q[i] = round(T, clamp((A[i]-Amin)*Δ⁻¹ + Tmin, Tmin, Tmax)) 
...

# Base.Array
Δ = (Qmax-Qmin)/(Tmax-Tmin)
A[i] = Qmin + (Q[i] - Tmin)*Δ
```

And I have two questions: 

1. Could I remove the `n::Integer` argument from `Base.Array` ? In the end, the range of values can be calculated  as `typemax(eltype(Q)) - typemin(eltype(Q))` instead of `2^n - 1`.
2. Does it make sense to create aliases for `LinQuantization` and `LinQuantArray`, since the alias is equally long as the original call? e.g. 
```julia 
LinQuantInt8Array(A::AbstractArray{T,N},dim::Int,e::Option{Tuple}) where {T,N} = LinQuantArray(Int8,A,dim,e)
``` 

Also  I am facing some issues. The back and forth conversion with  signed integers is not equal for some values, with a different of 1.0 or 2.0. For `Float16` and `Int16` some values of A2 end up being `-∞` or `∞`. I don't exactly know what the problem is.
```julia
for T in [Float64, Float32, Float16]
   for s in [(100,), (10,20), (13,14,15), (23,17,12,5)]:
        A = rand(T,s...)

        for U in [Int8, Int16,  Int32]
            A2 = Array{T}(LinQuantization(U,A))
            B = Array{T}(LinQuantization(U,A2))

            if A2 != B 
                println("T = ", T, " U = ", U)
            end 
            for i in eachindex(A2)
                if A2[i] != B[i]
                    println(" A2[i] = ", A2[i], " B[i] = ", B[i])
                end
            end
        end
    end
end

# T = Float32 U = Int32
#  A2[i] = 2.0507566e7 B[i] = 2.0507564e7
# T = Float32 U = Int32
#  A2[i] = 8.266649e6 B[i] = 8.266648e6
# T = Float32 U = Int32
#  A2[i] = 6.41308e6 B[i] = 6.413079e6
#  A2[i] = 848596.0 B[i] = 848595.0
#  A2[i] = 3.2166338e7 B[i] = 3.2166336e7
#  A2[i] = 1.717558e6 B[i] = 1.717557e6
#  A2[i] = 1.5504138e7 B[i] = 1.5504137e7
#  A2[i] = 3.065833e6 B[i] = 3.065832e6
#  A2[i] = 6.550137e6 B[i] = 6.550136e6
#  A2[i] = 9.974494e6 B[i] = 9.974493e6
# T = Float32 U = Int32
#  A2[i] = 1.3689715e7 B[i] = 1.3689714e7
#  A2[i] = 7.192029e6 B[i] = 7.192028e6
#  A2[i] = 8.769405e6 B[i] = 8.769404e6
#  A2[i] = 6.903743e6 B[i] = 6.903742e6
#  A2[i] = 2.544131e6 B[i] = 2.54413e6
#  A2[i] = 1.0867795e7 B[i] = 1.0867794e7
# ...
```

pabvald committed:
feat: support linear quantization for signed integers, support custom extrema

pabvald committed:
chore: up version to 0.3.0

pabvald committed:
undo: use Delta instead of alpha

pabvald committed:
rename: Maybe{T} to Option{T}

pabvald committed:
refactor: use eltype

pabvald commented:
TODOs: 
- [x] Fix issues
- [x] Extend tests
- [x] Extend README

milankl commented on Project.toml:
```diff
@@ -1,7 +1,7 @@
 uuid = "44354dd6-1ce2-4e80-a0fb-98a3a2b05088"
 authors = ["Milan <milankloewer@gmx.de> and contributors"]
-version = "0.2.1"
+version = "0.3.0"
```
If I'm not mistaken then we can implement this as a non-breaking change

milankl commented on src/LinLogQuantization.jl:
```diff
@@ -2,6 +2,8 @@ module LinLogQuantization
 
     export  LinQuantArray, LogQuantArray,
     LinQuant8Array, LinQuant16Array, LinQuant24Array, LinQuant32Array,
+    LinQuantUInt8Array, LinQuantUInt16Array, LinQuantUInt24Array, LinQuantUInt32Array,
+    LinQuantInt8Array, LinQuantInt16Array, LinQuantInt24Array, LinQuantInt32Array,
```
I think if we avoid defining new types then we can make this pull request non-breaking. So a `LinQuant8Array` of type `T <: Signed` would be what you're suggesting and `T <: Unsigned` is what we had before.

milankl commented on src/linquantarrays.jl:
```diff
@@ -10,84 +13,164 @@ Base.size(QA::LinQuantArray) = size(QA.A)
-    Amin = Float64(minimum(A))              # minimum of value range
-    Amax = Float64(maximum(A))              # maximum of value range
+    # range of values in A
+    Amin, Amax  = Float64(minimum(A)), Float64(maximum(A))    
```
I would use `extrema` from `Base` here. Otherwise the array `A` has to be scanned twice, `extrema` does both in one -- important for very large arrays and performance

milankl commented on src/linquantarrays.jl:
```diff
@@ -10,84 +13,164 @@ Base.size(QA::LinQuantArray) = size(QA.A)
+    A::AbstractArray,
+    extrema::Nothing
+) where {T<:Integer}
     all(isfinite.(A)) || throw(DomainError("Linear quantization only in (-∞,∞)"))
```
Looking at this here, we should probably first have the `extrema` call, then simply check whether min and max are finite (they wouldn't be if any element in `A` isn't finite). The problem with `all(isfinite.(A))` is that it'll allocate a bitarray and reduce that again to a single boolean. I know I wrote that, but it's not what we should do 😆

milankl commented on src/linquantarrays.jl:
```diff
@@ -10,84 +13,164 @@ Base.size(QA::LinQuantArray) = size(QA.A)
+function LinQuantization(
+    ::Type{T},
+    A::AbstractArray,
+    extrema::Nothing
```
I believe this should be `extrema::Nothing = nothing` so that the default uses the non-clamped version?

milankl commented on src/linquantarrays.jl:
```diff
@@ -10,84 +13,164 @@ Base.size(QA::LinQuantArray) = size(QA.A)
+function LinQuantization(
+    ::Type{T},
+    A::AbstractArray,
+    extrema::Tuple = extrema(A),
```
I think I'd prefer to turn this into a keyword argument?

pabvald replied:
Multiple dispatch doesn't support keyword arguments. If there is:

```julia 
function LinQuantization(
    ::Type{T},
    A::AbstractArray;
    extrema::Nothing=nothing
) where {T<:Integer}
...
end
```
and 
```julia 
function LinQuantization(
    ::Type{T},
    A::AbstractArray;
    extrema::Tuple
) where {T<:Integer}
...
end
```
the second definition will override the first one. To make it a keyword argument, we could do the following: 

```julia 
function LinQuantization(
    ::Type{T},
    A::AbstractArray;
    extrema::Option{Tuple}=nothing
) where {T<:Integer}
    # range of values in A
    Amin, Amax  = isnothing(extrema) ? Float64.(Base.extrema(A)) : Float64.(extrema)

    # guard against infinite values
    (isfinite(Amin) && isfinite(Amax)) || throw(DomainError("Linear quantization only in (-∞,∞)"))

    # minimum and maximum representable value of type T
    Tmin, Tmax = Float64(typemin(T)), Float64(typemax(T))

    # inverse spacing, set to zero for no range
    Δ⁻¹ = Amin == Amax ? zero(Float64) : (Tmax-Tmin)/(Amax-Amin)

    Q = similar(A,T)                        # preallocate

    if isnothing(extrema)
        # map minimum to typemin(T), maximum to typemax(t)
        @inbounds for i in eachindex(Q)
            Q[i] = round((A[i]-Amin)*Δ⁻¹ + Tmin)
        end
    else 
        # map minimum to typemin(T), maximum to typemax(t)
        # clamp to [Tmin,Tmax] removing out-of-range values
        @inbounds for i in eachindex(Q)
            Q[i] = round(T, clamp((A[i]-Amin)*Δ⁻¹ + Tmin, Tmin, Tmax))
        end
    end

    return LinQuantArray{T,ndims(Q)}(Q,Amin,Amax)
end
```

EDIT: fix typos

milankl commented on src/linquantarrays.jl:
```diff
@@ -10,84 +13,164 @@ Base.size(QA::LinQuantArray) = size(QA.A)
-    Qmin = Q.min                # min as Float64
-    Qmax = Q.max                # max as Float64
-    Δ = (Qmax-Qmin)/(2^n-1)     # linear spacing
+function Base.Array{U}(n::Integer, Q::LinQuantArray) where {U<:AbstractFloat}
```
yeah I think it makes sense to remove the `n` because that information is in the `eltype` of `Q`.

milankl commented on src/linquantarrays.jl:
```diff
@@ -10,84 +13,164 @@ Base.size(QA::LinQuantArray) = size(QA.A)
 Base.Array{T}(Q::LinQuantArray{UInt16,N}) where {T,N} = Array{T}(16,Q)
 Base.Array{T}(Q::LinQuantArray{UInt24,N}) where {T,N} = Array{T}(24,Q)
 Base.Array{T}(Q::LinQuantArray{UInt32,N}) where {T,N} = Array{T}(32,Q)
 
-Base.Array(Q::LinQuantArray{UInt8,N}) where N = Array{Float32}(8,Q)
-Base.Array(Q::LinQuantArray{UInt16,N}) where N = Array{Float32}(16,Q)
-Base.Array(Q::LinQuantArray{UInt24,N}) where N = Array{Float32}(24,Q)
-Base.Array(Q::LinQuantArray{UInt32,N}) where N = Array{Float64}(32,Q)
+# define default conversions for signed 8, 16, 24 and 32 bit
+Base.Array(Q::LinQuantArray{Int8,N}) where N = Array{Float32}(8,Q)
+Base.Array(Q::LinQuantArray{Int16,N}) where N = Array{Float32}(16,Q)
+Base.Array(Q::LinQuantArray{Int24,N}) where N = Array{Float32}(24,Q)
+Base.Array(Q::LinQuantArray{Int32,N}) where N = Array{Float64}(32,Q)
```
Once we take the `n` from the `eltype` in `Q` then these would only be needed to set a default floating-point type if not provided I guess!

milankl commented on src/linquantarrays.jl:
```diff
@@ -10,84 +13,164 @@ Base.size(QA::LinQuantArray) = size(QA.A)
 LinQuant32Array(A::AbstractArray{T,N},dim::Int) where {T,N} = LinQuantArray(UInt32,A,dim)
 
+# for unsigned integers  8,16,24 and 32 bit
+LinQuantUInt8Array(A::AbstractArray{T,N},dim::Int,e::Option{Tuple}) where {T,N} = LinQuantArray(UInt8,A,dim,e)
+LinQuantUInt16Array(A::AbstractArray{T,N},dim::Int,e::Option{Tuple}) where {T,N} = LinQuantArray(UInt16,A,dim,e)
+LinQuantUInt24Array(A::AbstractArray{T,N},dim::Int,e::Option{Tuple}) where {T,N} = LinQuantArray(UInt24,A,dim,e)
+LinQuantUInt32Array(A::AbstractArray{T,N},dim::Int,e::Option{Tuple}) where {T,N} = LinQuantArray(UInt32,A,dim,e)
+
+# for signed integers of 8,16,24 and 32 bit
+LinQuantInt8Array(A::AbstractArray{T,N},dim::Int,e::Option{Tuple}) where {T,N} = LinQuantArray(Int8,A,dim,e)
+LinQuantInt16Array(A::AbstractArray{T,N},dim::Int,e::Option{Tuple}) where {T,N} = LinQuantArray(Int16,A,dim,e)
+LinQuantInt24Array(A::AbstractArray{T,N},dim::Int,e::Option{Tuple}) where {T,N} = LinQuantArray(Int24,A,dim,e)
+LinQuantInt32Array(A::AbstractArray{T,N},dim::Int,e::Option{Tuple}) where {T,N} = LinQuantArray(Int32,A,dim,e)
```
We should be able to define these simply with the parameter `T` that `LinQuantArray{T}` already has. So like use `LinQuant8Array` for unsigned 8-bit integers, but if you want signed use `LinQuantArray{Int8}` that makes the unsigned integers appear as default but it's still convenient to write for signed integers without defining 8 new function names that are all exported and that basically just double up the type definition

milankl commented:
> Could I remove the `n::Integer` argument from `Base.Array` ? In the end, the range of values can be calculated  as `typemax(eltype(Q)) - typemin(eltype(Q))` instead of `2^n - 1`.

Yes, I agree with that. See comments.

> Does it make sense to create aliases for `LinQuantization` and `LinQuantArray`, since the alias is equally long as the original call? e.g.
> 
> 
> ```julia
> LinQuantInt8Array(A::AbstractArray{T,N},dim::Int,e::Option{Tuple}) where {T,N} = LinQuantArray(Int8,A,dim,e)
> ```

Yeah, no, I prefer not to export 8 new function names, the latter sounds better to me.


> Also I am facing some issues. The back and forth conversion with signed integers is not equal for some values

That's expected given the lossy compression. That information loss happens in the `round` function but it should be round-to-nearest tie-to-even but not sure we actually test that. Only for large integers that error becomes smaller but also depends on the range of values. We could formulate also a idempotency test because that loss would only happen on the first round trip.

> with a different of 1.0 or 2.0. 

Linear quantization introduces an absolute error that is bounded by $\Delta/2$ with $\Delta$ being the spacing between quants that's also computed for the (de-)quantization.

> For `Float16` and `Int16` some values of A2 end up being `-∞` or `∞`. I don't exactly know what the problem is.

Okay yeah that shouldn't happen, if inputs are finite, the quantization should create finite values too. But if you start in integer-quantization-space you can create numbers that are quantized perfectly representable but aren't in `Float16`.
E.g. you can represent -32768:32767 with Int16 (typemin-typemax range), but for a spacing of $\Delta = 2$ then your largest number you can represent becomes 65534 which is just larger than `floatmin(Float16)` which is 65504.

```julia
julia> Float16(typemax(Int16)*2)
Inf16
```

So finite floats should be quantized into finite (signed) integers, but vice versa can trigger overflow for floats!

pabvald committed:
refactor: implement feedback

pabvald commented:
> Okay yeah that shouldn't happen, if inputs are finite, the quantization should create finite values too. But if you start in integer-quantization-space you can create numbers that are quantized perfectly representable but aren't in Float16.

I have implemented your feedback but  I still have to check the issues with the overflowing of Float16.

Edit: Everything seems to be working now 😃

pabvald committed:
chore: rm unnecesary dependencies

pabvald committed:
refactor: make extrema a keyword argument

pabvald committed:
feat: export Int24

pabvald committed:
fix: signature of linear quantization along one dimension

pabvald committed:
feat: extend tests for signed integers

pabvald committed:
fix: set both extrema to positive values

pabvald committed:
undo: recover StatsBase dependency

milankl commented:
Fails because tuples aren't equal with different but otherwise equal `==` (not identical `===`) elements
```julia
   Expression: Base.extrema(A2) == ext
   Evaluated: (0.22f0, 0.75f0) == (0.22, 0.75)
```

pabvald committed:
chore: update README

pabvald committed:
chore: formatting

pabvald committed:
fix: make extrema keyword argument for LinQuantArray{U}()

pabvald committed:
tests: for custom extrema and negative values

pabvald commented:
@milankl I have updated the README, although the benchmarking is missing. Do you happen to still have the code for benchmarking?

pabvald committed:
fix: matrix for LogQuant cannot have negative values

pabvald commented:
Oops Copilot added negative values to the LogQuant tests

milankl commented:
I don't have the benchmarking anymore, I believe I did something like

```julia
julia> A = rand(10_000_000);   # 80MB vector
julia> sizeof(A)/1000^2
80.0

julia> using BenchmarkTools

julia> @btime LinQuant16Array($A);
  37.425 ms (7 allocations: 20.27 MiB)

julia> 80/40e-3    # Float64 -> UInt16 at 2000MB/s
2000.0
```

and coming from single precision

```julia
julia> A = rand(Float32, 10_000_000);

julia> @btime LinQuant16Array($A);
  42.537 ms (7 allocations: 20.27 MiB)

julia> 40/40e-3    # Float32 -> UInt16 at 1000MB/s
1000.0
```

milankl commented:
Happy to provide another review in the next days, just to check, the linear quantization for unsigned integers didn't change right? Because I'm very happy to have all your suggestions implemented but ideally with no breaking changes for the current linear quantization with unsigned integers. So just to check, how do you envisage a user (including you) to use the signed integer quantization?

pabvald commented:
It has complete backwards compatibility. I would like to extend the README with a use case about using signed integer quantization to quantize document embeddings for a RAG application but I didn't have the time. When dealing with document/word embeddings it makes sense to keep negative values and therefore using signed integer quantization.

pabvald committed:
test: to check backwards compatibility

pabvald committed:
docs: extend docstrings

pabvald commented:
@milankl When do you think you will be able to review it?

milankl commented:
This looks fantastic, thanks so much for all your work on this pull request. I'm on it, I'll do a review today/tomorrow. Found a few typos otherwise it looks like it's ready to go, but I'll review it more thoroughly!

milankl committed:
typos and linting

pabvald commented:
We are only missing the benchmarking for linear quantization with signed integers.

milankl commented on README.md:
```diff
@@ -172,28 +240,39 @@ With `dlin = b^(log_b(min) + dlog) - min` this can be transformed into
 
-24-bit quantisation is via `UInt24` from the `BitIntegers` package,
+
+### Signed integers
```
Do you still want to add benchmarking here? I believe signed/unsigned is not going to make a difference, but providing the extrema will, as that triggers the `clamp`

pabvald replied:
Yes, I would like to add it. I will benchmark linear quantization providing extrema then

milankl commented on src/linquantarrays.jl:
```diff
@@ -1,105 +1,187 @@
-"""Struct that holds the quantised array as UInts with an additional
-field for the min, max of the original range."""
-struct LinQuantArray{T,N} <: AbstractArray{Unsigned,N}
+const Option{T} = Union{T,Nothing}
```
Happy to define this but it's not strictly necessary?

pabvald replied:
No, it's not strictly necessary, just an alias I usually use : D

milankl commented on src/linquantarrays.jl:
```diff
@@ -1,105 +1,187 @@
+
+function LinQuantization(
+    ::Type{T},
+    A::AbstractArray;
```
wondering whether we strictly speaking need to constrain `AbstractArray{<:Real}` but I guess if it's complex then the conversion will fail anyway

pabvald replied:
yes, it wouldn't harm

milankl commented on src/linquantarrays.jl:
```diff
@@ -1,105 +1,187 @@
-# for 8,16,24 and 32 bit
+function LinQuantArray{U}(
+    A::AbstractArray{T,N},
+    dim::Int; 
```
I guess technically one would want for consistency to use a `dims` keyword, similar to how sorting along a dimension works
```julia
julia> sort(randn(5,5), dims=1)
5×5 Matrix{Float64}:
 -1.76503   -1.1454    -0.666234   -1.60678    -1.93496
 -1.09311   -0.555642  -0.374038   -1.01704    -1.10898
 -0.970622   0.138932  -0.353606    0.0905675  -0.414778
  0.505033   0.259424  -0.0457179   0.386296    0.0198439
  1.25263    1.52726    1.32485     1.93694     2.14597
```

but I'm happy either way for now, just though that `LinQuantArray(A, 2)` isn't as verbose as `LinQuantArray(A, dims=2)`

pabvald replied:
But it would be a breaking change, wouldn't it?

milankl replied:
it would be breaking once we agree on one way and decide to do the other, yes.

pabvald replied:
I find it better to use `dims` and be consistent with other Julia functions.

milankl commented on test/runtests.jl:
```diff
@@ -67,18 +176,84 @@ end
+        @test maximum(A) > 0
+
+        Q = LinQuantArray{Int32}(A, 4)
+        @test A ≈ Array{Float32}(Q)
```
if you test after the roundtrip then you can actually test for equality because rounding is an idempotent operation, introducing the rounding error only the first time

pabvald replied:
Actually I just copied your previous tests. I thought you might have a reason to so.

milankl replied:
Just added idempotency tests

milankl committed:
add some idempotency tests for along dimension

pabvald committed:
chore: benchmarking

milankl committed:
higher tolerance for Int8 in tests

milankl committed:
Merge branch 'linear-quantization-for-signed-integers' of https://github.com/pabvald/linlogquantization.jl into pr/pabvald/13

pabvald commented:
@milankl I have refactored the code to use a keyword argument `dims` instead of `dim`. Check it out and let me know if there's something that needs fixing. Otherwise, looking forward to merge this!

P.S. I have modified the version to `1.0.0` because it now includes breaking changes

pabvald commented:
@milankl Could you please run the test pipeline and merge it if everything is green 🙏🏻 ?

milankl commented:
Running now 🥳

pabvald commented:
@milankl There was a typo using the old notation without the keyword argument `dims`. Should work now.

milankl commented on Project.toml:
```diff
@@ -1,16 +1,16 @@
 uuid = "44354dd6-1ce2-4e80-a0fb-98a3a2b05088"
 authors = ["Milan <milankloewer@gmx.de> and contributors"]
-version = "0.2.1"
+version = "1.0.0"
```
Oh, mh. I see that we've added a new feature, but actually this is a non-breaking chance, so I don't necessarily want to go to v1.0 if it's not necessary?

pabvald replied:
The new notation requiring dims as keyword `dims = 4` is a breaking change, isn't it?

pabvald replied:
@milankl I suggest the following: 

1. I reset the branch to a commit previous to the standardisation of the `dims` argument.
2. You release 0.3.0 with the new support for signed integers.
3. I create a following PR with the standardisation of the `dims` argument to solve [Issue 14](https://github.com/milankl/LinLogQuantization.jl/issues/14)
4. You release 1.0.0 with the standard notation.

milankl replied:
I like the idea of releasing something non breaking (but that would be 0.2.2!) but if you want to also go straight to v1.0 I could be supportive of that even though I don't see the point. But given that you've put more effort than me into this repo I'd be happy to include you as a collaborator and then you have write access too and can take the post v1 development and maintaince in the direction you like.

But could you elaborate how the `dims` keyword is breaking? It's an optional keyword no?

pabvald replied:
> I like the idea of releasing something non breaking (but that would be 0.2.2!) 

mmm the most common versioning scheme is [Semantic Versioning (SemVer)](https://en.wikipedia.org/wiki/Software_versioning#Semantic_versioning), which uses a three-part number format: `MAJOR.MINOR.PATCH`. The support for unsigned integers is not a patch (a.k.a. bugfix), but anew functionality that doesn't break backwards compatibility, i.e. a _MINOR_. Am I missing something?

> But given that you've put more effort than me into this repo I'd be happy to include you as a collaborator and then you have write access too and can take the post v1 development and maintaince in the direction you like.

Becoming a collaborator would be great 😄 !

> But could you elaborate how the dims keyword is breaking? It's an optional keyword no?

Until now the signature of the function was 
```julia 
function LinQuantArray(::Type{TUInt},A::AbstractArray{T,N},dim::Int)
```
with `dim` being a positional argument. People could call the function in the two following ways: 
```julia-repl
julia> LinQuantArray(UInt8, A) # no dimension specified
``` 
or 
```julia-repl 
julia> LinQuantArray(UInt8, A, 4)
``` 

Using the standard **keyword** argument `dims`, the  signature 
```julia 
function LinQuantArray(
    ::Type{TInteger},
    A::AbstractArray{T,N};
    dims::Int,
    extrema::Option{Tuple} = nothing
)
```

People can call the function in the three following ways: 
```julia-repl
julia> LinQuantArray(UInt8, A) # no dimension specified
``` 
or
```julia-repl 
julia> LinQuantArray(UInt8, A, dims=4)
``` 
or
```julia-repl 
julia> LinQuantArray(UInt8, A; dims=4)
``` 

This means that previous calls will break and this is a breaking change.

milankl replied:
> > I like the idea of releasing something non breaking (but that would be 0.2.2!)
> 
> mmm the most common versioning scheme is [Semantic Versioning (SemVer)](https://en.wikipedia.org/wiki/Software_versioning#Semantic_versioning), which uses a three-part number format: `MAJOR.MINOR.PATCH`. The support for unsigned integers is not a patch (a.k.a. bugfix), but anew functionality that doesn't break backwards compatibility, i.e. a _MINOR_. Am I missing something?

This applies to v1.0 and later. Before v1 v0.1 -> v0.2 is considered a breaking change, and v0.1 -> v0.1.1 can be used for both bugfixes as well as new features. This is quite common because if there are other packages having a dependency on, say, v0.2 then v0.2.2 would be included in that compatibility. Otherwise people have to update their compat entries manually (or with compathelper) to v0.3 or v1.0. I'm happy to go to either of these to signal a new feature, but ideally not without also releasing v0.2.2 so that packages don't have to udpate.

pabvald replied:
Oh alright. Didn't know that. Thanks for the clarification. 

Then we can release: 
* `0.2.2` with the support for unsigned integers
*  `1.0.0` with the standard `dims` and and update of the Julia version and the dependencies

I will update this PR to leave the `dims` change out and create a different PR for that.

milankl commented on README.md:
```diff
@@ -172,28 +254,40 @@ With `dlin = b^(log_b(min) + dlog) - min` this can be transformed into
+| compression      | 1500 MB/s | 1500 MB/s | 1500 MB/s | 1500 MB/s |  100 MB/s |  100 MB/s | 1500 MB/s | 1500 MB/s |
+| decompression    | 16000 MB/s | 8000 MB/s | 16000 MB/s | 8000 MB/s | 8000 MB/s | 8000 MB/s | 16000 MB/s | 16000 MB/s |
+| **Custom extrema** |
+| compression      | 8000 MB/s | 8000 MB/s | 8000 MB/s | 8000 MB/s |    100 MB/s |    100 MB/s | 8000 MB/s | 8000 MB/s |
```
why are custom extrema ~5x faster to compress?

pabvald replied:
When extrema is `nothing`, the function calculates the extrema of the array `A` using `Base.extrema(A)`. This involves iterating over the entire array to find the minimum and maximum values, which can be time-consuming for large arrays?

milankl replied:
ouh yeah, I guess! Thanks for checking

milankl commented:
Shoot yeah, I forgot the renaming from `dim` to `dims`. I did put in the `dim` functionality but didn't use it much so intuitively didn't consider it to be public API. But yes, in that case can we go to v0.3 first, and then if you are arguing to go to v1 straight, happy to consider this too

pabvald committed:
format: files

pabvald committed:
chore: update Project.toml

pabvald committed:
chore: up version to 0.3.0

pabvald commented:
Version 0.3.0 is ready

pabvald committed:
fix: add missing signature LogQuantArray{U}

pabvald committed:
test, fix: use unsigned integers for logarithmic quantiztion tests

pabvald committed:
benchmark: overhead of Base.extrema

pabvald commented:
@milankl  The tests for logarithmic quantization were using signed integers, I guess I fixed it in a later commit. It should pass now.

I have also added benchmarking for `Base.extrema`.

milankl commented:
Awesome, so we merge this, `dim` is `dim` not `dims` yet. I would then actually just merge #15 too, and tag both as v0.3 denothing breaking and new features. Deciding on v1.0 we can then still do?
